### PR TITLE
[IR] Improve check for target extension types disallowed in globals.

### PR DESCRIPTION
**Pull request by jayfoad**
For target extension types that are not allowed to be used as globals,
also disallow them nested inside array and structure types.

### CODE REVIEW

File: llvm/lib/IR/Type.cpp [hunk 1]
```diff
@@ -425,6 +441,34 @@ bool StructType::isScalableTy(SmallPtrSetImpl<const Type *> &Visited) const {
...
+      return true;
+    }
+  }
+
+  // For structures that are opaque, return false but do not set the
+  // SCDB_NotContainsNonGlobalTargetExtType flag since it may gain non-global
+  // target extension types when it becomes non-opaque.
+  if (!isOpaque())
+    const_cast<StructType *>(this)->setSubclassData(
+        getSubclassData() | SCDB_NotContainsNonGlobalTargetExtType);
```
Comment (nikic) [inline llvm/lib/IR/Type.cpp:468]: :(

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/IR/Type.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/IR/Type.cpp:468]: Are you sad about the long name containing both "not" and "non"?

Comment (nikic) [reply llvm/lib/IR/Type.cpp:468]: I'm sad that our types aren't immutable and we have to do these contortions (https://discourse.llvm.org/t/recursive-types/82707).

File: llvm/lib/IR/Type.cpp [hunk 2]
```diff
@@ -903,7 +947,7 @@ static TargetTypeInfo getTargetTypeInfo(const TargetExtType *Ty) {
 
   // DirectX resources
   if (Name.starts_with("dx."))
-    return TargetTypeInfo(PointerType::get(C, 0));
+    return TargetTypeInfo(PointerType::get(C, 0), TargetExtType::CanBeGlobal);
```
Comment (jayfoad) [inline llvm/lib/IR/Type.cpp:950]: @bogner @hekota I had to do this to keep `check-clang-codegenhlsl` passing. Does it look right?

Comment (bogner) [reply llvm/lib/IR/Type.cpp:950]: Yes, this looks reasonable to me.

**Review by bogner — APPROVED** [review]

**Pull request merged**